File: python/src/advent_of_code/year2025/day7.py
from copy import copy
from time import time_ns

from advent_of_code.infra import read_input, report_times


def expand_node(node: tuple[int, int], space: list[str]) -> list[tuple[int, int]]:
    if space[node[0]][node[1]] != "^":
        return [(node[0] + 1, node[1])]

    return [(node[0] + 1, node[1] + 1), (node[0] + 1, node[1] - 1)]


def count_splits(space: list[str]) -> int:
    beams: set[int] = {space[0].index("S")}
    splits: int = 0
    for line in space[1:]:
        for index in copy(beams):
            if line[index] == "^":
                splits += 1
                beams.remove(index)
                beams.add(index + 1)
                beams.add(index - 1)
    return splits


def count_realities(
    beam_position: tuple[int, int],
    space: list[str],
    cache: dict[tuple[int, int], int],
) -> int:
    if beam_position in cache:
        return cache[beam_position]

    if beam_position[0] >= len(space):
        cache[beam_position] = 1
        return 1

    count = 0
    for node in expand_node(beam_position, space):
        count += count_realities(node, space, cache)
    cache[beam_position] = count
    return count


def day6() -> None:
    start = time_ns()

    # read and parse file
    inputs = read_input(2025, 7, example=False).splitlines()
    file_read_time = time_ns() - start

    # part 1
    print(count_splits(inputs))
    part1_time = time_ns() - start

    # part 2
    print(count_realities((0, inputs[0].index("S")), inputs, {}))
    part2_time = time_ns() - start

    # report times
    report_times(file_read_time, part1_time, part2_time)


if __name__ == "__main__":
    day6()

File: rust/year2025/src/bin/day7.rs
use std::collections::HashMap;
use std::error::Error;
use std::time::Instant;

use aoc_shared::{Map2D, read_input, report_times};

fn find_start(map: &Map2D<char>) -> (usize, usize) {
    for i in 0..map.width() {
        if *map.get(i, 0) == 'S' {
            return (i, 0);
        }
    }
    panic!("Could not find start.")
}

fn count_splits(start_position: (usize, usize), map: &Map2D<char>) -> usize {
    let mut active_beams: Vec<bool> = vec![false; map.width()];
    active_beams[start_position.0] = true;

    let mut splits: usize = 0;
    for y in 1..map.height() {
        for x in 0..map.width() {
            if *map.get(x, y) == '^' && active_beams[x] {
                splits += 1;
                active_beams[x] = false;
                active_beams[x + 1] = true;
                active_beams[x - 1] = true;
            }
        }
    }

    splits
}

fn count_realities(
    beam_position: (usize, usize),
    map: &Map2D<char>,
    cache: &mut HashMap<(usize, usize), usize>,
) -> usize {
    if let Some(&count) = cache.get(&beam_position) {
        return count;
    }

    if beam_position.1 >= map.height() - 1 {
        cache.insert(beam_position, 1);
        return 1;
    }

    let mut count: usize = 0;
    for node in expand_node(beam_position, map) {
        count += count_realities(node, map, cache);
    }
    cache.insert(beam_position, count);
    count
}

fn expand_node(node: (usize, usize), map: &Map2D<char>) -> Vec<(usize, usize)> {
    if *map.get(node.0, node.1) != '^' {
        return vec![(node.0, node.1 + 1)];
    }
    vec![(node.0 + 1, node.1 + 1), (node.0 - 1, node.1 + 1)]
}

fn main() -> Result<(), Box<dyn Error>> {
    let start = Instant::now();

    // read and parse file
    let input = read_input("2025", "7", false)?;
    let map: Map2D<char> = Map2D::from_lines(input);
    let start_position = find_start(&map);
    let file_read_time = start.elapsed();

    // part 1
    let part1_result: usize = count_splits(start_position, &map);
    let part1_time = start.elapsed();

    // part 2
    let mut cache: HashMap<(usize, usize), usize> = HashMap::new();
    let part2_result: usize = count_realities(start_position, &map, &mut cache);
    let part2_time = start.elapsed();

    // report results and times
    println!("{part1_result}");
    println!("{part2_result}");
    report_times(file_read_time, part1_time, part2_time);
    Ok(())
}
